kale/nbparser/parser.py

```diff
@@ -124,62 +116,88 @@ def parse_notebook(notebook):
+                functions_block.append(c.source)
+            if prev_step_name == 'pipeline-parameters':
+                pipeline_parameters.append(c.source)
+            # current_block might be None in case the first cells of the
+            # notebooks have not been tagged.
+            if prev_step_name:
+                # this notebook cell will be merged to a previous one that
+                # specified a step name
+                merge_code(nb_graph, prev_step_name, c.source)
         else:
-            current_block_global = False
-            current_block_pipeline_parameters = False
-            # TODO: Taking by default first block name tag. Need specific behavior
-            current_block = tags['block_names'][0]
-            for block_name in tags['block_names']:
-                # add node to DAG, adding tags and source code of notebook cell
-                if block_name not in nb_graph.nodes:
-                    _tags = copy.deepcopy(tags)
-                    _tags.block_name = block_name
-                    nb_graph.add_node(block_name, tags=_tags, source=c.source,
-                                      ins=set(), outs=set())
-                    if tags['previous_blocks']:
-                        for block in tags['previous_blocks']:
-                            nb_graph.add_edge(block, block_name)
-                else:
-                    merge_code(nb_graph, block_name, c.source)
-
-    # Prepend all global code blocks before every other one
-    if global_block != "":
-        for block in nb_graph:
-            block_data = nb_graph.nodes(data=True)[block]
-            nx.set_node_attributes(nb_graph, {block: {
-                'source': global_block + '\n\n' + block_data['source']}})
+            # add node to DAG, adding tags and source code of notebook cell
+            if step_name not in nb_graph.nodes:
+                nb_graph.add_node(step_name, source=[c.source],
+                                  ins=set(), outs=set())
+                for _prev_step_name in tags['prev_steps']:
+                    nb_graph.add_edge(_prev_step_name, step_name)
+            else:
+                merge_code(nb_graph, step_name, c.source)
+
+        prev_step_name = step_name
+
+    # Prepend any `imports` and `functions` cells to every Pipeline step
+    for step in nb_graph:
+        step_source = nb_graph.nodes(data=True)[step]['source']
+        nx.set_node_attributes(
+            nb_graph,
+            {step: {
+                'source': imports_block + functions_block + step_source}})
+
+    # merge together pipeline parameters
+    pipeline_parameters = '\n'.join(pipeline_parameters)
+
+    # make the nodes' code a single multiline string
+    # NOTICE: this is temporary, waiting for the artifacts-viz-feature
+    for step in nb_graph:
+        step_source = nb_graph.nodes(data=True)[step]['source']
+        nx.set_node_attributes(
+            nb_graph,
+            {step: {'source': '\n'.join(step_source)}})
```

Comment:
I'd prefer
```python
        step_dict = {step: {'source': '\n'.join(step_source)}}  # or any other variable name
        nx.set_node_attributes(nb_graph, step_dict)
```
Otherwise, the following is also fine
```python
        nx.set_node_attributes(nb_graph,                                         
                               {step: {'source': '\n'.join(step_source)}})
```